processing/src/main/java/io/druid/segment/filter/InFilter.java

```diff
@@ -69,40 +85,104 @@ public ImmutableBitmap apply(String value)
-            }
-          }
+          getPredicateFactory().makeStringPredicate()
       );
     }
   }
 
   @Override
   public ValueMatcher makeMatcher(ValueMatcherFactory factory)
   {
-    return factory.makeValueMatcher(
-        dimension, new Predicate<String>()
-        {
-          @Override
-          public boolean apply(String input)
+    return factory.makeValueMatcher(dimension, getPredicateFactory());
+  }
+
+  @Override
+  public boolean supportsBitmapIndex(BitmapIndexSelector selector)
+  {
+    return selector.getBitmapIndex(dimension) != null;
+  }
+
+  private DruidPredicateFactory getPredicateFactory()
+  {
+    return new DruidPredicateFactory()
+    {
+      @Override
+      public Predicate<String> makeStringPredicate()
+      {
+        if (extractionFn != null) {
+          return new Predicate<String>()
+          {
+            @Override
+            public boolean apply(@Nullable String input)
+            {
+              return values.contains(Strings.nullToEmpty(extractionFn.apply(input)));
+            }
+          };
+        } else {
+          return new Predicate<String>()
           {
-            if (extractionFn != null) {
-              input = extractionFn.apply(input);
+            @Override
+            public boolean apply(@Nullable String input)
+            {
+              return values.contains(Strings.nullToEmpty(input));
             }
-            return values.contains(Strings.nullToEmpty(input));
+          };
+        }
+      }
+
+      @Override
+      public DruidLongPredicate makeLongPredicate()
+      {
+        if (extractionFn != null) {
+          return new DruidLongPredicate()
+          {
+            @Override
+            public boolean applyLong(long input)
+            {
+              return values.contains(extractionFn.apply(input));
+            }
+          };
+        } else {
+          if (useLongHash) {
+            return new DruidLongPredicate()
+            {
+              @Override
+              public boolean applyLong(long input)
+              {
+                return longHashSet.contains(input);
+              }
+            };
+          } else {
+            return new DruidLongPredicate()
+            {
+              @Override
+              public boolean applyLong(long input)
+              {
+                return Arrays.binarySearch(longArray, input) >= 0;
+              }
+            };
           }
         }
-    );
+      }
+    };
   }
 
-  @Override
-  public boolean supportsBitmapIndex(BitmapIndexSelector selector)
+  private void setLongValues()
   {
-    return selector.getBitmapIndex(dimension) != null;
+    int idx = 0;
+    for (String value : values) {
+      Long longValue = Longs.tryParse(value);
+      if (longValue != null) {
+        if (useLongHash) {
+          longHashSet.add(longValue);
+        } else {
+          longArray[idx] = longValue;
```

Comment:
If any `values` are unparseable as longs, then this will leave zeroes behind in the `longArray`, which we don't want. The array needs to be shrunk if any `values` are unparseable.